### PR TITLE
storelimit: fix datarace from `getOrCreateStoreLimit`

### LINKED ISSUE #8253: storelimit: getOrCreateStoreLimit will meet datarace
### What did you do?

getOrCreateStoreLimit from different source

### What did you expect to see?

no datarace

### What did you see instead?

```
Write at 0x00c00070d928 by goroutine 270:
  github.com/tikv/pd/pkg/core/storelimit.(*limit).Reset()
      /home/lhy1024/pd/pkg/core/storelimit/store_limit.go:152 +0x265
  github.com/tikv/pd/pkg/core/storelimit.(*StoreRateLimit).Reset()
      /home/lhy1024/pd/pkg/core/storelimit/store_limit.go:126 +0x3b
  github.com/tikv/pd/pkg/core.(*StoresInfo).ResetStoreLimit.ResetStoreLimit.func1()
      /home/lhy1024/pd/pkg/core/store_option.go:258 +0x14b
  github.com/tikv/pd/pkg/core.(*StoreInfo).Clone()
      /home/lhy1024/pd/pkg/core/store.go:112 +0x17a
  github.com/tikv/pd/pkg/core.(*StoresInfo).ResetStoreLimit()
      /home/lhy1024/pd/pkg/core/store.go:870 +0x208
  github.com/tikv/pd/pkg/schedule/operator.(*Controller).getOrCreateStoreLimit()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller.go:989 +0x2b7
  github.com/tikv/pd/pkg/schedule/operator.(*Controller).ack()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller.go:576 +0x217
  github.com/tikv/pd/pkg/schedule/operator.(*Controller).removeOperatorInner()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller.go:650 +0x224
  github.com/tikv/pd/pkg/schedule/operator.(*Controller).addOperatorInner()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller.go:519 +0x35d
  github.com/tikv/pd/pkg/schedule/operator.(*Controller).PromoteWaitingOperator()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller.go:413 +0x356
  github.com/tikv/pd/pkg/schedule/operator.(*Controller).AddWaitingOperator()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller.go:345 +0xc3a
  github.com/tikv/pd/pkg/schedule.(*Coordinator).tryAddOperators()
      /home/lhy1024/pd/pkg/schedule/coordinator.go:444 +0x247
  github.com/tikv/pd/pkg/schedule.(*PatrolRegionContext).startPatrolRegionWorkers.func1()
      /home/lhy1024/pd/pkg/schedule/coordinator.go:216 +0x209

Previous read at 0x00c00070d928 by goroutine 272:
  github.com/tikv/pd/pkg/core/storelimit.(*StoreRateLimit).Rate()
      /home/lhy1024/pd/pkg/core/storelimit/store_limit.go:109 +0x217
  github.com/tikv/pd/pkg/schedule/operator.(*Controller).getOrCreateStoreLimit()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller.go:988 +0x111
  github.com/tikv/pd/pkg/schedule/operator.(*Controller).ack()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller.go:576 +0x217
  github.com/tikv/pd/pkg/schedule/operator.(*Controller).removeOperatorInner()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller.go:650 +0x224
  github.com/tikv/pd/pkg/schedule/operator.(*Controller).addOperatorInner()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller.go:519 +0x35d
  github.com/tikv/pd/pkg/schedule/operator.(*Controller).PromoteWaitingOperator()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller.go:413 +0x356
  github.com/tikv/pd/pkg/schedule/operator.(*Controller).AddWaitingOperator()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller.go:345 +0xc3a
  github.com/tikv/pd/pkg/schedule.(*Coordinator).tryAddOperators()
      /home/lhy1024/pd/pkg/schedule/coordinator.go:444 +0x247
  github.com/tikv/pd/pkg/schedule.(*PatrolRegionContext).startPatrolRegionWorkers.func1()
      /home/lhy1024/pd/pkg/schedule/coordinator.go:216 +0x209
```

### PR DESCRIPTION
**Pull request by lhy1024**
### What problem does this PR solve?

Issue Number: Close #8253

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test
- Manual test (add detailed scripts or steps below)

`go test -timeout 120s -run ^TestConcurrentAddOperatorAndSetStoreLimit$ github.com/tikv/pd/pkg/schedule/operator -race`
- run test `TestConcurrentAddOperatorAndSetStoreLimit` on master
```
WARNING: DATA RACE
Write at 0x00c0005ae480 by goroutine 220:
  github.com/tikv/pd/pkg/core/storelimit.(*limit).Reset()
      /home/lhy1024/pd/pkg/core/storelimit/store_limit.go:151 +0x227
  github.com/tikv/pd/pkg/core/storelimit.(*StoreRateLimit).Reset()
      /home/lhy1024/pd/pkg/core/storelimit/store_limit.go:126 +0x3b
  github.com/tikv/pd/pkg/core.(*StoresInfo).ResetStoreLimit.ResetStoreLimit.func1()
      /home/lhy1024/pd/pkg/core/store_option.go:258 +0x14b
  github.com/tikv/pd/pkg/core.(*StoreInfo).Clone()
      /home/lhy1024/pd/pkg/core/store.go:112 +0x17a
  github.com/tikv/pd/pkg/core.(*StoresInfo).ResetStoreLimit()
      /home/lhy1024/pd/pkg/core/store.go:870 +0x208
  github.com/tikv/pd/pkg/schedule/operator.(*Controller).getOrCreateStoreLimit()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller.go:989 +0x2b7
  github.com/tikv/pd/pkg/schedule/operator.(*Controller).ExceedStoreLimit()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller.go:966 +0x464
  github.com/tikv/pd/pkg/schedule/operator.(*Controller).AddOperator()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller.go:355 +0x6b
  github.com/tikv/pd/pkg/schedule/operator.TestConcurrentAddOperatorAndSetStoreLimit.func1()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller_test.go:987 +0x2fb
  github.com/tikv/pd/pkg/schedule/operator.TestConcurrentAddOperatorAndSetStoreLimit.gowrap1()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller_test.go:994 +0x41

Previous read at 0x00c0005ae480 by goroutine 221:
  github.com/tikv/pd/pkg/core/storelimit.(*limit).Available()
      /home/lhy1024/pd/pkg/core/storelimit/store_limit.go:162 +0xb2
  github.com/tikv/pd/pkg/core/storelimit.(*StoreRateLimit).Available()
      /home/lhy1024/pd/pkg/core/storelimit/store_limit.go:101 +0x3a
  github.com/tikv/pd/pkg/schedule/operator.(*Controller).ExceedStoreLimit()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller.go:970 +0x4e5
  github.com/tikv/pd/pkg/schedule/operator.(*Controller).AddOperator()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller.go:355 +0x6b
  github.com/tikv/pd/pkg/schedule/operator.TestConcurrentAddOperatorAndSetStoreLimit.func1()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller_test.go:987 +0x2fb
  github.com/tikv/pd/pkg/schedule/operator.TestConcurrentAddOperatorAndSetStoreLimit.gowrap1()
      /home/lhy1024/pd/pkg/schedule/operator/operator_controller_test.go:994 +0x41
```
- run test `TestConcurrentAddOperatorAndSetStoreLimit` with this PR
```
ok  	github.com/tikv/pd/pkg/schedule/operator	1.473s
```
### Release note

```release-note
None.
```

### CODE REVIEW

File: pkg/core/storelimit/store_limit.go [hunk 1]
```diff
@@ -128,12 +129,15 @@ func (l *StoreRateLimit) Reset(rate float64, typ Type) {
...
-	limiter    *ratelimit.RateLimiter
-	ratePerSec float64
+	limiter         *ratelimit.RateLimiter
+	ratePerSecMutex syncutil.RWMutex
+	ratePerSec      float64
 }
 
 // Reset resets the rate limit.
 func (l *limit) Reset(ratePerSec float64) {
+	l.ratePerSecMutex.Lock()
```
Comment (rleungx) [inline pkg/core/storelimit/store_limit.go:139]: Will it happen in the current code or just in your testing code?

Comment (lhy1024) [reply pkg/core/storelimit/store_limit.go:139]: I think datarace can happen in the current code, we call `getOrCreateStoreLimit` in many places now, like in `AddOperator` or `tryAddOperators`, where we call `ExceedStoreLimit` to check the limit.

If the `getOrCreateStoreLimit` enters the `ResetStoreLimit` branch, then datarace is possible.

Comment (lhy1024) [reply pkg/core/storelimit/store_limit.go:139]: And just like the test I built, this scenario is also possible to datarace if we set storelimit when other code calls `getOrCreateStoreLimit`.

Comment (rleungx) [reply pkg/core/storelimit/store_limit.go:139]: seems it is caused by https://github.com/tikv/pd/pull/8032

Comment (lhy1024) [reply pkg/core/storelimit/store_limit.go:139]: > seems it is caused by #8032

sure, this test will fail since #8032

**Review by rleungx — APPROVED** [review]

**Review by nolouch — APPROVED** [review]

**Pull request merged**